File: VBNetCore.Mvvm/ViewModels/PropertyChangedBase.vb
Imports System.ComponentModel
Imports System.Linq.Expressions
Imports System.Runtime.CompilerServices
Imports System.Windows.Input
Imports VBNetCore.Mvvm.Helpers

Namespace VBNetCore.Mvvm.ViewModels
    ''' <summary>
    ''' Base class implementing <seecref="INotifyPropertyChanged"/>.
    ''' </summary>
    Public MustInherit Class PropertyChangedBase
        Implements INotifyPropertyChanged
        ''' <summary>
        ''' Occurs when a property value changes.
        ''' </summary>
        Public Event PropertyChanged As PropertyChangedEventHandler Implements INotifyPropertyChanged.PropertyChanged

        ''' <summary>
        ''' Storage for fields
        ''' </summary>
        Protected _properties As Dictionary(Of String, Object) = New Dictionary(Of String, Object)()

        ''' <summary>
        ''' Notifies clients that all properties may have changed.
        ''' </summary>
        ''' <remarks>
        ''' This method raises the <seecref="PropertyChanged"/> event with <seecref="String.Empty"/> as the property name.
        ''' </remarks>
        Protected Sub Refresh()
            Call CommandManager.InvalidateRequerySuggested()
            Me.OnPropertyChanged(String.Empty)
        End Sub

        ''' <summary>
        ''' Raises the <seecref="PropertyChanged"/> event.
        ''' </summary>
        ''' <paramname="propertyName">The name of the changed property.</param>
        Protected Overridable Sub OnPropertyChanged(ByVal Optional propertyName As String = Nothing)
            RaiseEvent PropertyChanged(Me, New PropertyChangedEventArgs(propertyName))
        End Sub

        ''' <summary>
        ''' Raises the <seecref="PropertyChanged"/> event.
        ''' </summary>
        ''' <paramname="property">The changed property.</param>
        ''' <typeparamname="TProperty">The type of the changed property.</typeparam>
        Protected Sub OnPropertyChanged(Of TProperty)(ByVal [property] As Expression(Of Func(Of TProperty)))
            Me.OnPropertyChanged(GetMemberName([property]))
        End Sub

        ''' <summary>
        ''' Set field value and Raises the <seecref="PropertyChanged"/> event.
        ''' </summary>
        ''' <typeparamname="TProperty"></typeparam>
        ''' <paramname="value"></param>
        ''' <paramname="propertyName"></param>
        ''' <returns></returns>
        Protected Overridable Function SetProperty(Of TProperty)(ByVal value As TProperty, <CallerMemberName> ByVal Optional propertyName As String = Nothing) As Boolean
            If EqualityComparer(Of TProperty).Default.Equals(GetProperty(Of TProperty)(propertyName), value) Then Return False
            If Not String.IsNullOrEmpty(propertyName) Then
                _properties(propertyName) = value
                OnPropertyChanged(propertyName)

                Return True
            Else
                Return False
            End If
        End Function

        ''' <summary>
        ''' Set field value and Raises the <seecref="PropertyChanged"/> event. 
        ''' </summary>
        ''' <typeparamname="TProperty"></typeparam>
        ''' <paramname="value"></param>
        ''' <paramname="property"></param>
        ''' <returns></returns>
        Protected Overridable Function SetProperty(Of TProperty)(ByVal value As TProperty, ByVal [property] As Expression(Of Func(Of TProperty))) As Boolean
            Return SetProperty(Of TProperty)(value, GetMemberName([property]))
        End Function

        ''' <summary>
        ''' Gets the value of a property
        ''' </summary>
        ''' <typeparamname="TProperty"></typeparam>
        ''' <paramname="propertyName"></param>
        ''' <returns></returns>
        Protected Overridable Function GetProperty(Of TProperty)(<CallerMemberName> ByVal Optional propertyName As String = Nothing) As TProperty
            If Not String.IsNullOrEmpty(propertyName) Then
                Dim value As Object = Nothing
                If _properties.TryGetValue(propertyName, value) Then Return If(value Is Nothing, Nothing, CType(value, TProperty))
            End If
            Return Nothing
        End Function

        ''' <summary>
        ''' Gets the value of a property
        ''' </summary>
        ''' <typeparamname="TProperty"></typeparam>
        ''' <paramname="property"></param>
        ''' <returns></returns>
        Protected Overridable Function GetProperty(Of TProperty)(ByVal [property] As Expression(Of Func(Of TProperty))) As TProperty
            Return GetProperty(Of TProperty)(GetMemberName([property]))
        End Function

    End Class
End Namespace

File: VBNetCore.Mvvm/ViewModels/PropertyChangedBase.cs
using NetCore.Mvvm.Helpers;
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Linq.Expressions;
using System.Runtime.CompilerServices;
using System.Windows.Input;

namespace NetCore.Mvvm.ViewModels
{
    /// <summary>
    /// Base class implementing <see cref="INotifyPropertyChanged"/>.
    /// </summary>
    public abstract class PropertyChangedBase : INotifyPropertyChanged
    {
        /// <summary>
        /// Occurs when a property value changes.
        /// </summary>
        public event PropertyChangedEventHandler? PropertyChanged;

        /// <summary>
        /// Storage for fields
        /// </summary>
        protected Dictionary<string, object> _properties = new Dictionary<string, object>();

        /// <summary>
        /// Notifies clients that all properties may have changed.
        /// </summary>
        /// <remarks>
        /// This method raises the <see cref="PropertyChanged"/> event with <see cref="string.Empty"/> as the property name.
        /// </remarks>
        protected void Refresh()
        {
            CommandManager.InvalidateRequerySuggested();
            OnPropertyChanged(string.Empty);
        }

        /// <summary>
        /// Raises the <see cref="PropertyChanged"/> event.
        /// </summary>
        /// <param name="propertyName">The name of the changed property.</param>
        protected virtual void OnPropertyChanged(string? propertyName = null)
        {
            PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(propertyName));
        }

        /// <summary>
        /// Raises the <see cref="PropertyChanged"/> event.
        /// </summary>
        /// <param name="property">The changed property.</param>
        /// <typeparam name="TProperty">The type of the changed property.</typeparam>
        protected void OnPropertyChanged<TProperty>(Expression<Func<TProperty>> property)
        {
            OnPropertyChanged(ExpressionHelper.GetMemberName(property));
        }

        /// <summary>
        /// Set field value and Raises the <see cref="PropertyChanged"/> event.
        /// </summary>
        /// <typeparam name="TProperty"></typeparam>
        /// <param name="value"></param>
        /// <param name="propertyName"></param>
        /// <returns></returns>
        protected virtual bool SetProperty<TProperty>(TProperty value, [CallerMemberName] string? propertyName = null)
        {
            if (EqualityComparer<TProperty>.Default.Equals(GetProperty<TProperty>(propertyName), value)) return false;
            if (!string.IsNullOrEmpty(propertyName))
            {
                _properties[propertyName] = value;
                OnPropertyChanged(propertyName);

                return true;
            }
            else
            {
                return false;
            }
        }

        /// <summary>
        /// Set field value and Raises the <see cref="PropertyChanged"/> event. 
        /// </summary>
        /// <typeparam name="TProperty"></typeparam>
        /// <param name="value"></param>
        /// <param name="property"></param>
        /// <returns></returns>
        protected virtual bool SetProperty<TProperty>(TProperty value, Expression<Func<TProperty>> property)
        {
            return SetProperty<TProperty>(value, ExpressionHelper.GetMemberName(property));
        }

        /// <summary>
        /// Gets the value of a property
        /// </summary>
        /// <typeparam name="TProperty"></typeparam>
        /// <param name="propertyName"></param>
        /// <returns></returns>
        protected virtual TProperty GetProperty<TProperty>([CallerMemberName] string? propertyName = null)
        {
            if (!string.IsNullOrEmpty(propertyName))
            {
                object? value = null;
                if (_properties.TryGetValue(propertyName, out value))
                    return value == null ? default(TProperty) : (TProperty)value;
            }
            return default(TProperty);
        }

        /// <summary>
        /// Gets the value of a property
        /// </summary>
        /// <typeparam name="TProperty"></typeparam>
        /// <param name="property"></param>
        /// <returns></returns>
        protected virtual TProperty? GetProperty<TProperty>(Expression<Func<TProperty>> property)
        {
            return GetProperty<TProperty>(ExpressionHelper.GetMemberName(property));
        }

    }
}
